src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -379,7 +378,15 @@ void PipelineExecutor::initializeExecution(size_t num_threads, bool concurrency_
 
 void PipelineExecutor::spawnThreads()
 {
-    while (auto slot = cpu_slots->tryAcquire())
+    std::unique_lock lock(spawn_lock, std::try_to_lock);
+    if (!lock.owns_lock())
+    {
+        /// Someone is already spawning threads, skip.
+        return;
+    }
+
+    AcquiredSlotPtr slot;
+    while (tasks.shouldSpawn() && (slot = cpu_slots->tryAcquire()))
```

Comment:
could we maybe use `thread_num` from l.391? would be nice to avoid taking a lock (but `shouldSpawn` currently does that)

Reply:
`thread_num` is not the idle threads count. To avoid lock, maybe we can use a atoimc `idle_count` and updated if `threads_queue` size changed.